api/envoy/api/v2/discovery.proto

```diff
@@ -93,3 +93,92 @@ message DiscoveryResponse {
   // required for non-stream based xDS implementations.
   string nonce = 5;
 }
+
+// IncrementalDiscoveryRequest and IncrementalDiscoveryResponse are used in a
+// new GRPC endpoint for Incremental XDS. The feature is not supported for REST
+// management servers.
+//
+// With Incremental XDS, the IncrementalDiscoveryResponses do not need to
+// include a full snapshot of the tracked resources. Instead
+// IncrementalDiscoveryResponses are a diff to the state of a XDS client.
+// In incremental XDS there are per resource versions which allows to track
+// state at the resource granularity.
+// An XDS Incremental session is always in the context of a GRPC bidirectional
+// stream. This allows the XDS server to keep track of the state of XDS clients
+// connected to it.
+//
+// In Incremental XDS the nonce field is required and used to pair
+// IncrementalDiscoveryResponse to a IncrementalDiscoveryRequest ACK or NACK.
+// A message level system_version_info is present for debugging purposes only.
+//
+// IncrementalDiscoveryRequest can be sent in 3 situations:
+//   1. Initial message in a XDS bidirectional GRPC stream.
+//   2. As a ACK or NACK response to a previous IncrementalDiscoveryResponse.
+//      In this case the response_nonce is set to the nonce value in the Response.
+//      ACK or NACK is determined by the absence or presence of error_detail.
+//      In the case of a ACK the system_version_info must match the value in the
+//      associated Response. Otherwise it must be set the value of the most recently
+//      ACKed Response.
+//   3. Spontaneous IncrementalDiscoveryRequest from the client.
+//      This can be done to dynamically add or remove elements from the tracked
+//      resource_names set.
+message IncrementalDiscoveryRequest {
+  // The system_version_info provided in the request messages will be the
+  // system_version_info received with the most recent successfully processed
+  // response or empty on the first request.
+  string system_version_info = 1;
+
+  // The node making the request.
+  core.Node node = 2;
+
+  // Type of the resource that is being requested, e.g.
+  // "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment". This is implicit
+  // in requests made via singleton xDS APIs such as CDS, LDS, etc. but is
+  // required for ADS.
+  string type_url = 3;
+
+  // A list of Resource names to add to the list of tracked resources.
+  // The empty list is a subscription to all resources.
+  repeated string resource_names = 4;
+  // A list of Resource names to remove from the list of tracked resources.
+  repeated string resource_names_unsubscribe = 5;
+
+  // This map must be populated when the IncrementalDiscoveryRequest is the
+  // first in a stream. The keys are know resources names to the CDS client and
+  // the value is the correcponding resource level version info.
+  resource_versions map<string,string> = 6;
```

Comment:
typos: known and corresponding. CDS should be XDS, I think.

Reply:
I think the type/field name are also swapped here.

Reply:
Re "I think the type/field name are also swapped here.": I'm not sure what you mean here. My intention is to have keys be resource names and values be their version.